### ISSUE TITLE
Bluetooth SPP support

### ISSUE THREAD

**Issue by JOHugo** [rkl099/Appinventor-SerialOTG#6]
Hi. Would love to get Bluetooth SPP support. The serial port of my Arduino is connected to a HC-05 Bluetooth module. In windows I can pair with the HC-05 and it looks like a normal serial port. I can even upload my HEX code from the Arduino IDE to my Arduino over Bluetooth. With AI2 I can use the Bluetooth blocks to access my Arduino, but it would be nice to integrate it with the HEX upload feature in your addon. Will this be a possibility for future releases?

**Comment by rkl099** [comment]
It sounds interesting to be able to use Bluetooth for serial communication (and HEX file upload). There is support for this in the Physicaloid library, but I have not tested it yet. I will order a HC05 and do some testing to see if I can make it work with AI2.

**Comment by JOHugo** [comment]
Up to now I've been using this cheaper HC-05 compatible Bluetooth module. It works fine with both 3.3V and 5V. 
https://www.aliexpress.com/item/33012208387.html
Here is the datasheet for the module. Their AT commands are also easier to use than the original HC-05
http://myosuploads3.banggood.com/products/20190129/20190129043725SKUA87502.pdf
This article explains how to connect the module + get auto reset on the DTR pin of an Arduino PRO mini working
https://forum.arduino.cc/t/bluetooth-upload-to-pro-mini-working/336021
Here is a nice video showing how to change the baud rate and use the module
https://www.youtube.com/watch?v=HxjE9l9xcSs

In my case I use Optiboot that defaults to 115200 Baud. The Bluetooth module defaults to 9600 Baud. The only thing you need to do is to change the Baud rate to 115200. Connect to the module at 9600 and do the following.
AT
AT+BAUD8    (use BAUD7 for 57600 Baud)
AT+RESET
From now on it will do serial passthrough at 115200 Baud. 
Pair it with windows. Get the serial port number assigned by Windows. Set you Arduino IDE com port and you can upload your code (Programmer type = ArduinoISP). When Windows connects to the module, the DTR line will cause the Arduino to reset and the upload will begin. Disconnect the DTR line for normal serial over Bluetooth SPP use.

**Comment by rkl099** [comment]
I have examined the library that I have modified and used for my extensions. It had a Bluetooth serial class, but it only implements BT server (slave), so the external device has to initialize the connection. I think this is possible with HC-05, (not with HC-06) but this does not seems to be so useful.

**Comment by JOHugo** [comment]
Thanks for the feedback and for having a look.

**Comment by rkl099** [comment]
Now I have rewritten part of the library to get the server function to work.
I can see the use for reprogramming Arduino over Bluetooth, so I implemented a client function as well! . I hope that you can test the reprogramming and see if its usefull.
You will find the extension on https://github.com/rkl099/Appinventor-SerialOTG_Debug

Please return with comments.

**Comment by JOHugo** [comment]
I've compiled your extension and I can send commands and receive data from
my hardware. I was hoping to make use of the Physicaloid extension to
upload a HEX file to the Arduino, but I just found out that the Physicaloid
extension does not include the firmware upload code. I'm more of a hardware
guy and a bit of a novice with coding and I am not sure if I can contribute
much. I was wondering if it would be possible to make use of the Physicaloid
java library functions and just pass it the HEX file.

https://github.com/ksksue/PhysicaloidLibrary

Upload a firmware from Android to Arduino:
Physicaloid mPhysicaloid = new Physicaloid(this);
mPhysicaloid.upload(Boards.ARDUINO_UNO, "/sdcard/arduino/Blink.hex");

**Comment by rkl099** [comment]
I have updated the library and include a better test program with basic communication, SerialOTG, BT client and server functions. The upload function is available in the extension and in the test program for a limited number of boards, but many boards  use the same parameters. I have also included the library so you can write your own programs as requested. See read.md (in debug directory.
I can see that the upload function sends data over BT SPP, but I dont have the hardware setup now to verify it completely, but the function ha previously been verified over serialOTG.

**Comment by JOHugo** [comment]
Wow !!!, I've managed to upload new Arduino HEX code to my board, over
Bluetooth, using your test program (works 50% of the time).

I've got embedded hardware, with an atmega328p chip + optiboot loader
installed. The Pro328_5V board type works just fine (screenshot of my
settings is attached). For programming, I use the MCU-INT (EN on some BT
modules) signal to generate an Arduino reset when a BT connection opens
(attached diagram). I've inserted some code from your "Open" function into
the "Upload" function so that a BT connection will open + reset the CPU,
just before the upload happens.

To get a successful upload, I have to close+restart the app from new. If I
try an upload after playing around with some of the functions, then it
fails most of the time.
The other thing that leads to a failed upload, is when the BT connection
takes slightly longer to open (visible via the blue led), just before the
upload happens. Looks like the timing between an Arduino reset and code
upload time is very critical.

It would be nice to get some visual feedback of upload progress.

I'll try to test an upload to an UNO over the weekend.

**Comment by rkl099** [comment]
Nice to hear!
Could you try to do an upload with the BT connection closed. It should not be necessary to have the connection open, since the upload function makes an open() itself. The timing can be critical with the reset from the HC-05 and I have noticed there is a difference in the time to do an open. You may have to do an open-close first before the upload.
The upload is done in an background process, but I will add code to report the progress, as well as status for the connection. 
Good luck with your testing!

**Comment by rkl099** [comment]
I think you always have to close the connection before an upload. If it's open then the upload function uses the open connection and you will not have a reset signal from the HC-05.

**Comment by JOHugo** [comment]
I've tested uploading many times now. I can see the upload code opens the
BT connection at the beginning. An open-close before doing the upload seems
to help, but I get an successful upload only once in a while. The rest of
the time it fails and corrupts the optiboot loader as well. Command line
uploads with AVRdude over BT (Win 10) works every time. I see AVTdude was
ported to Android at some point, but I don't see any recent activity on
github. https://github.com/syslogic/avrdude-jni

**Comment by rkl099** [comment]
For me it sounds like data corruption. I will continue to go through the code and add some more status indications. I will also have a test setup with a HC-05 soon.
One question. How do you power the arduino? If you power it via USB from a phone/tablet then will the usb-serial converter be active. This can affect the data.

**Comment by JOHugo** [comment]
How do you power the arduino?
It's an embedded board with a switch-mode 5v regulator, powered by a 12v
alarm battery. No cables connected to the phone.
The HEX file is quite big = 87,367 bytes.
I'm busy trying to set up a serial monitor so that I can capture the data
received by the Bluetooth module and write it to a file on my computer.
Is it possible to write a copy of the serial data sent over bluetooth
during an upload to a file? I can then compare the data that was sent, with
the data that was received.

**Comment by JOHugo** [comment]
A couple of questions:
- what's the default setting for the serial stop bit? Is it set?
- What do you use RTS for? Does it have any influence with BT comms?
- what's the purpose of the "0" "1" next to "Open"?

**Comment by rkl099** [comment]
0,1 is for the testprogram led.ini. Just sends "0"  "1" to turn led off on.

RTS DTR have no meaning for BT, only for serial (for reset and may be needed for the USB-Serial IC)
Parity=NONE, Databits=8, StopBits=1 (Baud depends on board), but I don't think they have any meaning for BT.
They are all do nothing in the BT protocol

**Comment by rkl099** [comment]
Do you have the possibility to test the upload function with direct serial connection?

**Comment by JOHugo** [comment]
I think with the correct cable in my phone it would be possible. I'll see
if I can get one.

**Comment by rkl099** [comment]
Which processor are you using? Is it included in the board types I have in the app?

**Comment by JOHugo** [comment]
I got hold of an OTG converter and my USB serial adapter works fine with an
Android serial terminal app. My serial adapter has a PL2303HX chip inside,
so it should work with SerialOTG. What should I change in
SerialOTG_uppl.aia to download over serial, instead of bluetooth ?

**Comment by rkl099** [comment]
You only have to select Ser OTG instead of BT client before uploading. Test to open once and get open status (first time you open, you have to give permission, next time you do the actual open), then you know that the serial adapter is detected.  You can also check the communication by connecting rx to tx on the adapter. If you send some data from the app, the you should receive the same in the line under.

**Comment by JOHugo** [comment]
RX/TX loopback works fine. If I type 3 + "send", then I receive a number 3.
RX seems to work and I get the normal text from my arduino after a reset.
Now the weird stuff, if I send command no 3 to the arduino (type 3 +
"send"), then the arduino doesn't recognise it.
Now if I use the "Serial USB terminal" android app, with the same cable
attached, then the arduino recognises the command 3 (type 3 + "send") and
it sends back data.
Both "Serial USB terminal" and "SerialOTG_upp" are configured for 115200
Baud ,8 bits, parity none, and one stop bit.

but my board doesn't respond to commands that I send to it.
Commands+response works fine with the same cable, using the "Serial USB
terminal" android app.

**Comment by JOHugo** [comment]
RX/TX loopback works fine. If I type 3 + "send", then I receive a number 3.
RX seems to work and I get the normal text from my arduino after a reset.
Now the weird stuff, if I send command no 3 to the arduino (type 3 +
"send"), then the arduino doesn't recognise it.
Now if I use the "Serial USB terminal" android app, with the same cable
attached, then the arduino recognises the command 3 (type 3 + "send") and
it sends back data.
Both "Serial USB terminal" and "SerialOTG_upp" are configured for 115200
Baud ,8 bits, parity none, and one stop bit.

Might be a CR+LF at the end

**Comment by rkl099** [comment]
OK, the app sends the text without CR LF.  No CR or LF. Do you need CR LF?

**Comment by JOHugo** [comment]
Not critical. My arduino code reads serial data, until it gets a CR LF.

**Comment by rkl099** [comment]
OK, If you wait for CR LF, then you never gets the data.
I think that serial terminal adds CR, LF as default

**Comment by JOHugo** [comment]
Ok, RTS does not reset my board correctly. I first need to do some
soldering to sort it out before I can run the upload tests.

**Comment by rkl099** [comment]
The upload code  sets RTS/DTR low for 50mS then high for 50mS. Then it "drain" incoming characters for 1S before starting the upload.
I think its best to open the serial connection and set RTS DTR high before trying to upload.

**Comment by JOHugo** [comment]
My  PL2303HX serial adapter is giving the same problem from my PC. I've
swapped it with another serial adapter with a FT232RL chip. This one is
working with both AVRdude on my PC and with SerialOTG. Upload with
SerialOTG works 10/10 times, but it's 10x slower than an upload with
AVRdude. At least it works every time.

**Comment by rkl099** [comment]
Then we know that the upload code is ok. I will have a look if I can find a reason for the slow upload!
I'm also working on to get status information from the upload process.

**Comment by JOHugo** [comment]
Just for info:
Upload via Avrdude = 25 sec
Upload via SerialOTG = 2min 30 sec

**Comment by rkl099** [comment]
I have a new version of the library and the app for testing. on SerialOTG_Debug.
It has display of file error and upload progress in nr of pages.
I have tested it on an Arduino Uno with serial and BT upload of (short) programs without any problems. I have found no reason for the long upload times that you mentioned. Maybe many retries?
If you are familiar with reading log files with Logcat Reader, then maybe you can send me some logfiles.

**Comment by JOHugo** [comment]
I found the problem. I'm using Optiboot https://github.com/Optiboot/optiboot
for my boot loader and it's not compatible with any of the supported
boards. I've now installed the UNO's bootloader (think it's derived from
Optiboot) and I selected UNO as my board type. Bluetooth downloads are
working perfectly now, without the long delays :-)
One comment. The upload progress does not reset when I do another hex file
upload.
Thanks for your great work. I'm very excited about it.
